### PR TITLE
Display results of ApplicantFilter in SingleSearchRequestTable

### LINKED ISSUE #3648: Update SingleSearchRequestTable to also accept applicantFilter
Related: https://github.com/GCTC-NTGC/gc-digital-talent/issues/3016

## Description
Currently, the `<SingleSearchRequestTable />` component is only setup to accept a `poolCandidateFilter`. We need to refactor it to accept either a `poolCandidateFilter` OR an `applicantFilter` so it can show the appropriate table data.  This will require creating another `searchPoolCandidates` query that accepts an `ApplicantFilterInput` in the where clause. Also, in SingleSearchRequest.tsx (where  `<SingleSearchRequestTable />` is called) make sure to filter data from applicantFilter instead of poolCandidateFilter if possible.

If a PoolCandidateSearchRequest contains an applicantFilter instead of a poolCandidateFilter, the request page must use the 2.0 version of the User table, with extra filters and updated style. It must also use the same same search query as the talent-search page 2.0.

## Screenshot: Prototype
[![a screenshot of ...][image]][image-link]
<sup>[Click screenshot to visit Figma.][image-link]</sup>

## Acceptance Criteria
- [ ] SingleSearchRequestTableApi accepts a `poolCandidateFilter` OR an `applicantFilter`
- [ ] If the request has an applicantFilter, the page shows the same users which would have been returned on the Talent Search 2.0 page
    - [ ] Links to the Admin User page with the tabs
    - [ ] Has the nice filters from User Table 2.0
    - [ ] Allows for downloading of user profiles.
- [ ] If a request has a poolCandidateFilter and no applicantFilter, it can be unchanged.

   [image]: https://user-images.githubusercontent.com/4313717/174098348-c22a9852-c598-4ea0-be9f-6837a10df666.png
   [image-link]: https://www.figma.com/proto/XS4Ag6GWcgdq2dBlLzBkay/Admin-v3-(Profile-release)?node-id=410%3A1079&starting-point-node-id=410%3A1984
   [figma]: https://go.talent.c4nada.ca/figma
   [sitemap]: https://go.talent.c4nada.ca/sitemap

### PR DESCRIPTION
**Pull request by tristan-orourke**
Resolves #3648.

This should be the minimal viable solution to #3648, ie allowing admins to review the list of applicants who meet the requirements of an ApplicantFilter request. We can spin off any non-essential features of #3648 into other tickets.

Testing:
1. ensure ApplicantSearch feature flag is true
2. Go to /search, and submit a request that results in at least some candidates.
3. Sign in to the admin dashboard. View the most recent request.
4. The table of users at the bottom should have the same number of rows as the number of candidates according to the search page.
5. Each row should have a link to the Users page, with the tabs
6. The filters in the table should be editable (without modifying the original request!), but if they're not, make that a new issue.

- [x] e567fbe rename `transform*` functions https://github.com/GCTC-NTGC/gc-digital-talent/pull/4623#issuecomment-1307431675
- [x] link ticket to improve seeders  from https://github.com/GCTC-NTGC/gc-digital-talent/pull/4623#issuecomment-1307456587
  - https://github.com/GCTC-NTGC/gc-digital-talent/issues/4357#issuecomment-1307639764
- [x] create ticket for cypress tests from https://github.com/GCTC-NTGC/gc-digital-talent/pull/4623#issuecomment-1307456587
  - https://github.com/GCTC-NTGC/gc-digital-talent/issues/4666

### CODE REVIEW

**Review by yonikid15 — COMMENTED** [review]
It looks like it's coming together nicely 🎉 

I checked the network requests along with the DB and it looks like it's working as expected, switching seamlessly btw `ApplicantFilterInput` and the soon to be deprecated `PoolCandidateFilterInput`. Appears to be showing the correct Users. One thing we need to improve is seeding the request tables with Users. Most of the time they are empty. However, that's out of scope for this issue.

I'll let you finish up the rest (cleaing up linting errors, removing commented out code, any other refactoring) before a final review. Nice work!

**Comment by vd1992** [comment]
Am I missing a step
codegen -> dev -> open up a request at `/talent-requests/:id`

I don't have any users load
image

image

**Comment by vd1992** [comment]
think something might be weird with my request page since it lets me send that request with non-zero candidates/applicants but SQL querying for just pool and active gives zero results

image

**Comment by tristan-orourke** [comment]
Made it possible to initialize the filters which appear in a UserTable filter dialog, and used the request.applicantFilter to initialize it!

@vd1992 there are a few extant bugs which might explain why you see non-zero results on the search page, and none on the admin SingleSearchRequest table. #4557 was already on the board, and I just discovered #4660!

**Comment by tristan-orourke** [comment]
@yonikid15 the searchApplicants query was a leftover from previous implementation - no need to test it if I can delete it! 😝

**Comment by petertgiles** [comment]
Don't forget your translations. :smile: 
```
{
  "hmacO5": {
    "defaultMessage": "Request doesn't include a filter!",
    "description": "Null state for a request not including a filter."
  }
}
```

File: frontend/admin/src/js/components/user/UserTableFilterDialog.tsx [hunk 1]
```diff
@@ -257,18 +257,21 @@ export type UserTableFiltersProps = Pick<
...
+  initialActiveFilters,
   ...rest
 }: UserTableFiltersProps) => {
   const [isOpen, setOpen] = React.useState<boolean>(isOpenDefault);
   const { emptyFormValues } = useFilterOptions(enableEducationType);
-  const [activeFilters, setActiveFilters] =
-    useState<FormValues>(emptyFormValues);
+  const [activeFilters, setActiveFilters] = useState<FormValues>(
+    initialActiveFilters ?? emptyFormValues,
+  );
```
Comment (patcon) [inline frontend/admin/src/js/components/user/UserTableFilterDialog.tsx:274]: Just to hint at "initialState", which is var name in useState docs that we pass in:

```suggestion
  const initialFiltersState = activeFilters ?? emptyFormValues;
  const [activeFilters, setActiveFilters] = useState<FormValues>(initialFiltersState);
```

File: frontend/admin/src/js/components/user/UserTableFilterDialog.tsx [hunk 2]
```diff
@@ -257,18 +257,21 @@ export type UserTableFiltersProps = Pick<
   "onSubmit" | "enableEducationType"
 > & {
   isOpenDefault?: boolean;
+  initialActiveFilters?: FormValues;
```
Comment (patcon) [inline frontend/admin/src/js/components/user/UserTableFilterDialog.tsx:260]:
```suggestion
  activeFilters?: FormValues;
```

File: frontend/admin/src/js/components/user/UserTable.tsx [hunk 3]
```diff
@@ -129,41 +201,7 @@ export const UserTable: React.FC = () => {
...
-      isProfileComplete: data.profileComplete[0] ? true : undefined,
-      jobLookingStatus: data.jobLookingStatus.map((status) => {
-        return stringToEnumJobLooking(status);
-      }),
-      poolFilters: data.pools.map((pool) => {
-        const poolString = pool;
-        return { poolId: poolString };
-      }),
-    });
+    setUserFilterInput(formValuesToUserFilterInput(data));
```
Comment (patcon) [inline frontend/admin/src/js/components/user/UserTable.tsx:204]: Starting functions with verbs?

```suggestion
    setUserFilterInput(transformUserFormValuesToFilterInput(data));
```

File: frontend/admin/src/js/components/user/UserTable.tsx [hunk 4]
```diff
@@ -45,6 +45,74 @@ import {
...
+      return stringToEnumJobLooking(status);
+    }),
+    poolFilters: data.pools.map((pool) => {
+      const poolString = pool;
+      return { poolId: poolString };
+    }),
+  };
+}
+
+function userFilterInputToFormValues(
```
Comment (patcon) [inline frontend/admin/src/js/components/user/UserTable.tsx:86]:
```suggestion
function transformUserFormValuesToFilterInput(
```

**Review by petertgiles — CHANGES_REQUESTED** [review]
On my initial test, I expected to get one result but I ended up getting two.  If it's helpful I can dig in and try to find the discrepancy.

image

File: frontend/admin/src/js/components/user/UserTable.tsx [hunk 5]
```diff
@@ -80,13 +148,17 @@ const emailLinkAccessor = (email: string | null, intl: IntlShape) => {
...
+};
+export const UserTable = ({ initialFilterInput }: UserTableProps) => {
   const intl = useIntl();
   const paths = useAdminRoutes();
   const { pathname } = useLocation();
 
-  const [userFilterInput, setUserFilterInput] = useState<UserFilterInput>();
+  const [userFilterInput, setUserFilterInput] = useState<UserFilterInput>(
+    initialFilterInput ?? {},
+  );
```
Comment (patcon) [inline frontend/admin/src/js/components/user/UserTable.tsx:161]:
```suggestion
  const initialStateFilterInput = initialFilterInput ?? {};
  const [userFilterInput, setUserFilterInput] = useState<UserFilterInput>(initialStateFilterInput);
```

File: frontend/admin/src/js/components/searchRequest/SingleSearchRequestTable.tsx [hunk 6]
```diff
@@ -247,19 +258,165 @@ export const SingleSearchRequestTable: React.FunctionComponent<
...
+export function checkIsLegacyFilter(
+  filter: AbstractFilter,
+): filter is PoolCandidateFilter {
+  // eslint-disable-next-line no-underscore-dangle
+  if (filter.__typename === "PoolCandidateFilter") return true;
+
+  return false;
+}
+
+const transformFilterToInput = (
```
Comment (patcon) [inline frontend/admin/src/js/components/searchRequest/SingleSearchRequestTable.tsx:272]:
```suggestion
const transformPoolCandidateFilterToFilterInput = (
```

File: frontend/admin/src/js/components/searchRequest/SingleSearchRequestTable.tsx [hunk 7]
```diff
@@ -247,19 +258,165 @@ export const SingleSearchRequestTable: React.FunctionComponent<
...
-  poolCandidateFilter: PoolCandidateFilterInput;
-}> = ({ poolCandidateFilter }) => {
-  const [result] = useSearchPoolCandidatesQuery({
-    variables: { poolCandidateFilter },
+  filter: AbstractFilter;
+}> = ({ filter }) => {
+  const isLegacyFilter = checkIsLegacyFilter(filter);
+
+  const poolCandidateFilterInput = isLegacyFilter
+    ? transformFilterToInput(filter)
```
Comment (patcon) [inline frontend/admin/src/js/components/searchRequest/SingleSearchRequestTable.tsx:403]:
```suggestion
    ? transformPoolCandidateFilterToFilterInput(filter)
```

File: frontend/admin/src/js/components/searchRequest/SingleSearchRequestTable.tsx [hunk 8]
```diff
@@ -247,19 +258,165 @@ export const SingleSearchRequestTable: React.FunctionComponent<
...
+}
+
+// Maps each property in ApplicantFilterInput to a function which transforms the matching value from an ApplicantFilter object to the appropriate shape for ApplicantFilterInput.
+type MappingType = {
+  [Property in keyof ApplicantFilterInput]: (
+    x: ApplicantFilter[Property],
+  ) => ApplicantFilterInput[Property];
+};
+
+const transformApplicantFilterToApplicantFilterInput = (
```
Comment (patcon) [inline frontend/admin/src/js/components/searchRequest/SingleSearchRequestTable.tsx:340]:
```suggestion
const transformApplicantFilterToFilterInput = (
```

**Comment by tristan-orourke** [comment]
@petertgiles I love the translation check! This line is the kind of error message we should never expect to ever be seen, so I just put it through Deepl.

File: frontend/admin/src/js/components/searchRequest/SingleSearchRequest.tsx [hunk 9]
```diff
@@ -211,49 +209,7 @@ export const SingleSearchRequest: React.FunctionComponent<
     searchRequest;
   // TODO: data filter data from applicantFilter instead of poolCandidateFilter if possible.
```
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply frontend/admin/src/js/components/searchRequest/SingleSearchRequest.tsx:210]: Yes it can!

**Comment by yonikid15** [comment]
> @yonikid15 the searchApplicants query was a leftover from previous implementation - no need to test it if I can delete it! 😝

Haha ya I forgot to remove that part of my comment. My bad 😆.

**Comment by patcon** [comment]
Some naming suggestions that basically follow this recco:

no more things in this format, and instead use => this:
- `fooToFooInput` => `transformXFooToFooInput` (use verb and explicit X)
- `fooInputToFoo` => `transformXFooInputToFoo` (use verb and explicit X)
- `transformXFooToXFooInput` => `transformXFooToFooInput` (X implied same)
- but keep this fully explicit: `transformXFooToYFoo` => explicitly change type

Considered (as discarded) idea of suggesting a different verb for different change types, like:
- `transform` for changes between input and non-input: `transformXFooToFooInput` or `transformXFooInputToFoo`
- `transpose` for changes between types: `tranposeXFooToYFoo`

Seems too premature, but the sort of thinking we could create language for as we extract these things into named functions :)

**Comment by tristan-orourke** [comment]
@petertgiles whether its fixed in this PR or another, please take a look into that discrepancy 🙏 . But to make sure you're not wasting your time, first check if its caused by https://github.com/GCTC-NTGC/gc-digital-talent/issues/4557 or https://github.com/GCTC-NTGC/gc-digital-talent/issues/4660.

**Comment by patcon** [comment]
#todo create ticket to improve seeders

**Comment by petertgiles** [comment]
> whether its fixed in this PR or another, please take a look into that discrepancy

OK, I did the analysis. In this case, the discrepancy was from missing the "would accept temporay" issue that Matt is working on in  #4659.

I did notice some other concerning things though:
- The search page is sending a request for a given number of _pool candidates_ but the admin request page is showing a list of _users_ that match the filters which would be the same for a single pool, usually, I think, but it's weird.
- The requests page also had the "is profile complete" logic.  In theory a user couldn't have a pool candidate match returned if their profile wasn't yet complete but it's just weird.

**Comment by petertgiles** [comment]
By the way, not sure why the Actions aren't running but this branch isn't building right now for me.
image

**Review by petertgiles — APPROVED** [review]
In my testing I'm continuing to get discrepancies but they're all around language ability which was fixed in #4660 (but I don't think this branch includes) and the wouldAcceptTemporary fixed in #4659 but is not yet merged.  There's a lot of noise but I think this works.  :smile:

**Comment by petertgiles** [comment]
> Frankly, using UserTable right now is a shortcut, just so we can deploy ASAP.

Fair enough.  Is that needed work represented on the board somewhere?

**Comment by tristan-orourke** [comment]
Made a ticket for changing it to a list of PoolCandidates #4672

**Pull request merged**